power_grid_model_c/power_grid_model/include/power_grid_model/topology.hpp

```diff
@@ -633,6 +634,35 @@ class Topology {
             {.component_obj_idx = comp_topo_.power_sensor_object_idx, .objects_coupling = comp_coup_.node},
             comp_coup_.power_sensor,
             [this](Idx i) { return comp_topo_.power_sensor_terminal_type[i] == MeasuredTerminalType::node; });
+
+        // branch 'from' current sensors
+        // include all branch3 sensors
+        auto const predicate_from_current_sensor = [this](Idx i) {
+            using enum MeasuredTerminalType;
+
+            return comp_topo_.current_sensor_terminal_type[i] == branch_from ||
+                   // all branch3 sensors are at from side in the mathemtical model
+                   comp_topo_.current_sensor_terminal_type[i] == branch3_1 ||
+                   comp_topo_.current_sensor_terminal_type[i] == branch3_2 ||
+                   comp_topo_.current_sensor_terminal_type[i] == branch3_3;
+        };
+        SensorBranchObjectFinder const object_finder_from_current_sensor{
+            .sensor_obj_idx = comp_topo_.current_sensor_object_idx,
+            .sensor_terminal_type = comp_topo_.current_sensor_terminal_type,
+            .branch_coupling = comp_coup_.branch,
+            .branch3_coupling = comp_coup_.branch3};
+
+        // branch 'from' current sensors
+        couple_object_components<&MathModelTopology::n_branch>(
+            [](MathModelTopology& math_topo) -> auto& { return math_topo.current_sensors_per_branch_from; },
+            object_finder_from_current_sensor, comp_coup_.current_sensor, predicate_from_current_sensor);
+
+        // branch 'to' current sensors
+        couple_object_components<&MathModelTopology::n_branch>(
+            [](MathModelTopology& math_topo) -> auto& { return math_topo.current_sensors_per_branch_to; },
+            {.component_obj_idx = comp_topo_.current_sensor_object_idx, .objects_coupling = comp_coup_.branch},
+            comp_coup_.current_sensor,
+            [this](Idx i) { return comp_topo_.current_sensor_terminal_type[i] == MeasuredTerminalType::branch_to; });
```

Comment:
here and elsewhere: there is some duplication with power sensors. i've reduced some using `power_or_current_sensor_c` but there still some improvements that can be made. if you have some ideas on how and when, please LMK